### DISCUSSION TITLE
Can MolToSmiles force the SMILES to start&end at a particular atom??

### DISCUSSION THREAD

**Discussion by Haoxu11225** [rdkit/rdkit#7199] [Q&A]
We know that MolToSmiles can force SMILES to start at a particular atom，like：
         -----------------------------------------------------------------------------------
         mol = Chem.MolFromSmiles('Oc1ccc(cc1)C(=O)')
         SMILES=Chem.MolToSmiles(mol, rootedAtAtom=1, canonical=False)
         Out:c1(O)ccc(C=O)cc1
         --------------------------------------------------------------------------------------
         But whether MolToSmiles has any way of specifying the last atom of SMILES, or both?
         Or is there any other way to force the position of the atoms in the generated SMILES, but still represent the original chemical structure?
          If you have any suggestions, thank you for leaving a message

**Comment by greglandrum** [comment]
@Haoxu11225 there's not currently any way to do this and it's not clear why you would want to specify the last atom in a SMILES.
What are you trying to accomplish?

**Comment by e-dorigatti** [comment]
@greglandrum this would definitely come useful when using chemical language models to generate molecules starting from a known fragment. For this, one would need a SMILES string whose last atom corresponds to the attachment point from where the model would complete the rest of the molecule. See this paper for an example: https://chemrxiv.org/engage/chemrxiv/article-details/65e718eee9ebbb4db9f21886

**Reply by greglandrum** [reply]
Without reading the preprint (don't have time for that), there's a much, much easier way to do this which has the added advantage of allowing you to have multiple attachment points: just use fragments in the SMILES and really high ring-closure digit numbers.
Here's an example:
```
In [4]: m = Chem.MolFromSmiles('c1nc%(100)nc%(200)c1.C%(100)C(=O)O.F%(200)')

In [5]: Chem.MolToSmiles(m)
Out[5]: 'O=C(O)Cc1nccc(F)n1'
```
Andrew Dalke has elaborated upon this idea:
https://sourceforge.net/p/rdkit/mailman/message/37253976/

**Reply by Haoxu11225** [reply]
e-dorigatti understood my question very well, because high degree of freedom controllable string normalization is very important for chemical language models. Since you can specify the starting atom. I wonder if it would be difficult to specify the last atom, or both, according to your rules. If one day you have an update in this area, it will certainly add a very important feature to RDKIT.

**Reply by shienliu** [reply]
```
def assiggn_last_atom_of_smiles(mol, lastAtomType):
    for _ in range(50):
        smi = Chem.MolToSmiles(mol, doRandom=True)
        #print(smi)
        m = Chem.MolFromSmiles(smi)
        length = len(mol.GetAtoms())
        if m.GetAtomWithIdx(length-1).GetSymbol() == lastAtomType:
            print(smi)
            break

m = Chem.MolFromSmiles('c1nc%(100)nc%(200)c1.C%(100)C(=O)O.Cl%(200)')
assiggn_last_atom_of_smiles(m, 'O')
```
I don't know if this code helps.
